File: api/src/services/personas.py
"""Persona pool for conversation sessions.

Each session randomly selects agent_count personas from PERSONA_POOL.
The persona name is used as the text-label convention ("Alice: ...") that
the single ADK agent uses to identify the active speaker.
"""

from __future__ import annotations

import random
from dataclasses import dataclass


@dataclass(frozen=True)
class Persona:
    name: str        # stored in Firestore, used as speaker_id
    label: str       # used in text-label convention ("Alice:", "Bob:", ...)
    personality: str # injected into the agent's session context


PERSONA_POOL: list[Persona] = [
    Persona(
        name="alice",
        label="Alice",
        personality=(
            "warm and curious; always asks at least one genuine follow-up question "
            "per turn and shows enthusiasm for what the user says"
        ),
    ),
    Persona(
        name="bob",
        label="Bob",
        personality=(
            "thoughtful and slightly skeptical; offers gentle counterpoints and "
            "alternative perspectives, but stays friendly"
        ),
    ),
    Persona(
        name="emma",
        label="Emma",
        personality=(
            "energetic and enthusiastic; loves sharing short personal anecdotes "
            "to relate to the topic and uses exclamations naturally"
        ),
    ),
    Persona(
        name="david",
        label="David",
        personality=(
            "calm and analytical; takes a moment to think, then gives concise "
            "well-reasoned observations; rarely interrupts"
        ),
    ),
    Persona(
        name="mia",
        label="Mia",
        personality=(
            "playful and witty; uses light humour and casual language to keep "
            "the mood relaxed, but stays on topic"
        ),
    ),
]

_POOL_BY_NAME: dict[str, Persona] = {p.name: p for p in PERSONA_POOL}


def select_personas(count: int) -> list[Persona]:
    """Randomly select `count` distinct personas from the pool."""
    count = max(1, min(count, len(PERSONA_POOL)))
    return random.sample(PERSONA_POOL, count)


def get_persona(name: str) -> Persona | None:
    return _POOL_BY_NAME.get(name)

File: frontend/src/personas.ts
// Mirrors api/src/services/personas.py's PERSONA_POOL. The backend randomly
// selects `agent_count` of these per session; the frontend needs the same
// catalog so it can render a consistent color/label for whichever persona
// name comes back in session.participants / speakerId, without the backend
// having to send display metadata over the wire.

export type PersonaMeta = { name: string; label: string; personality: string; personalityJa: string; color: string };

export const PERSONA_POOL: PersonaMeta[] = [
  { name: "alice", label: "Alice", personality: "warm and curious, always asks a follow-up", personalityJa: "温かく好奇心旺盛。いつも一言質問を返す", color: "var(--color-persona-alice)" },
  { name: "bob", label: "Bob", personality: "thoughtful, offers gentle counterpoints", personalityJa: "思慮深く、やんわり反対意見も出す", color: "var(--color-persona-bob)" },
  { name: "emma", label: "Emma", personality: "energetic, loves sharing personal stories", personalityJa: "エネルギッシュで、自分の体験談を話すのが好き", color: "var(--color-persona-emma)" },
  { name: "david", label: "David", personality: "calm and analytical, concise and rarely interrupts", personalityJa: "冷静で分析的。簡潔で、あまり口を挟まない", color: "var(--color-persona-david)" },
  { name: "mia", label: "Mia", personality: "playful and witty, keeps the mood relaxed", personalityJa: "遊び心があってウィットに富み、場を和ませる", color: "var(--color-persona-mia)" },
];

const BY_NAME = new Map(PERSONA_POOL.map((persona) => [persona.name, persona]));

export function personaFor(name: string): PersonaMeta {
  return BY_NAME.get(name) ?? { name, label: name.slice(0, 1).toUpperCase() + name.slice(1), personality: "", personalityJa: "", color: "var(--color-muted)" };
}
